### ISSUE TITLE
specify rules for 'block' decoration

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#1123]
Block decoration:
- only applies to a struct type
- A block-decorated struct must not contain another block-decorated struct at any nesting level

What's it used to represent?
- block-decorated struct is used for UBO, SSBO.  A block-decorated struct for these maps 1-1 to a binding slot.
    - Later I expect we'll have arrays of descriptors.  In this case the element of those arrays will be block-decorated, still with the 1-1 correspondence with a binding slot.

In SPIR-V / Vulkan they can also be used to collect builtin variables for input/output interface.  That has seen mixed use in Vulkan as far as I can tell.    That is not needed for WGSL, so I won't advocate for it.

**Comment by m-schuetz** [comment]
Would it be possible to allow defining structs without the need to specify offsets? 

E.g., I've got this struct here, and if I remove the offset decorator, chrome throws an error. However, it can be quite cumbersome to reevaluate offsets manually when changing structs. It also adds some additional mental overhead while reading the code.

```
[[block]] struct Batch {
	[[offset(0)]] chunk_start : u32;
	[[offset(4)]] chunk_size : u32;
	[[offset(8)]] points_read : u32;
	[[offset(12)]] dec_value : u32;
	[[offset(16)]] dec_length : u32;
	[[offset(20)]] dec_offset : u32;
};
```
nicer:
```
[[block]] struct Batch {
	chunk_start : u32;
	chunk_size : u32;
	points_read : u32;
	dec_value : u32;
	dec_length : u32;
	dec_offset : u32;
};
```

but:

```Structure id 8 decorated as Block must be explicitly laid out with Offset decorations```

**Comment by dj2** [comment]
This has been an open point of discussion. The explicit offsets force the programmer to think about them, and makes it very easy to map back to the offsets used when writing the data on the API side. For simple cases like the above, they are a bit of a pain, but for things like `mat2x2` where the offset is different from what would be expected in GLSL, having the user explicitly state them makes it clearer what's happening.

On the flip side, they are annoying to type/update. There maybe work on the compiler level to spit out what we think are the correct offsets, so then it's a copy/paste matter to update the script file.

Basically, it's a safety vs convenience tradeoff. Providing them is more work, but the shader author is telling us exactly what they mean and we aren't guessing.

**Comment by m-schuetz** [comment]
I understand. On a similar note, I've found it also a bit cumbersome to specify an array of a primitive type as input. For example, I'd like to define:

```
[[binding(0), set(0)]] var<storage_buffer> source : array<u32>;

// access via 
source[i];

```

which doesn't compile. Instead, it seems like I currently have to use 

```
[[block]] struct Source {
	[[offset(0)]] values : [[stride(4)]] array<u32>;
};

[[binding(0), set(0)]] var<storage_buffer> source : Source;

...

// access via
source.values[i]


```

Would it be possible to allow the short version? It looks like the long version only adds stride information, which could also be integrated into the short version?




Chrome: ```Variables identified with the StorageBuffer storage class are used to access transparent buffer backed resources. Such variables must be typed as OpTypeStruct, or an array of this type```

**Comment by dj2** [comment]
There are two ways to define arrays, runtime arrays and non-runtime. The difference is if you provide the array size. `array<u32, 5>` is an array of 5 elements of u32 and can be used most places. An `array<u32>` is a runtime sized array and _must_ be the last member of a block decorated structure due to rules in how runtime arrays work.

**Comment by m-schuetz** [comment]
> An array<u32> is a runtime sized array and must be the last member of a block decorated structure due to rules in how runtime arrays work.

The short version would just provide syntactic sugar that reduces the size of the storage buffer definition to 25% of the lines, has the same amount of information in it, and can be compiled to the same thing. It would make it faster to write, understand, and maintain wgsl code.

**Comment by dj2** [comment]
We're opting for explicit for the MVP. Yes, it's more lines, but for now making things more explicit that can be relaxed latter is a better option than relaxing things now which we can't make restrictive later.

This is the kind of sugar we could consider post MVP, but I don't see it making it for MVP.

**Comment by dneto0** [comment]
The rules for [[block]] are specified, certainly by the time #1524 landed.
Also #2104 is proposed to remove the attribute entirely.
